src/compiler/crystal/repl/context.cr

```diff
@@ -0,0 +1,495 @@
+require "./repl"
+
+# This class contains all of the global data used to interpret a single
+# program. For example, it includes the memory region to store constants
+# and class variables, what are all the know symbols, and a few more things.
+class Crystal::Repl::Context
+  record MultidispatchKey, obj_type : Type, call_signature : CallSignature
+
+  getter program : Program
+
+  # A hash of Def to their compiled representation, so we don't compile
+  # a single method multiple times.
+  # The exceptions are methods with non-captured blocks.
+  getter defs : Hash(Def, CompiledDef)
+
+  # Information about all known constants.
+  getter! constants : Constants
+
+  # Information about all known class variables.
+  getter! class_vars : ClassVars
+
+  # libffi information about external functions.
+  getter lib_functions : Hash(External, LibFunction)
+
+  # Cache of multidispatch expansions.
+  getter multidispatchs : Hash(MultidispatchKey, Def)
+
+  # The single closure function that we use for all function pointers (procs)
+  # passed to C. This closure function knows all the information about
+  # the callback being passed, and is able to run the code associated to it.
+  getter ffi_closure_fun : LibFFI::ClosureFun
+
+  # The memory where constants are stored. Refer to `Constants` for more on this.
+  property constants_memory : Pointer(UInt8)
+
+  # The memory where class vars are stored. Refer to `ClassVars` for more on this.
+  property class_vars_memory : Pointer(UInt8)
+
+  # Associated an FFI::Closure's code to a CompiledDef.
+  # When we set an extern struct's field that is a Proc, we create
+  # an FFI::Closure object and set that instead of a Proc.
+  # In case a user reads that field back we need to create a Proc,
+  # we can't use an FFI::Closure object. In that case we lookup
+  # the proc in this Hash.
+  getter ffi_closure_to_compiled_def : Hash(Void*, CompiledDef)
+
+  @pkg_config_path : String?
+
+  def initialize(@program : Program)
+    @program.flags << "interpreted"
+
+    @gc_references = [] of Void*
+
+    @defs = {} of Def => CompiledDef
+    @defs.compare_by_identity
+
+    @lib_functions = {} of External => LibFunction
+    @lib_functions.compare_by_identity
+
+    @dl_handles = {} of LibType => Void*
+
+    @symbol_to_index = {} of String => Int32
+    @symbols = [] of String
+
+    @multidispatchs = {} of MultidispatchKey => Def
+
+    @ffi_closure_contexts = {} of {UInt64, UInt64} => FFIClosureContext
+
+    # TODO: finish porting all of LibM instrinsics
```

Comment:
Oops, it's a dead comment now. Those intrinsics were done in a different way. Feel free to remove it!

Reply:
There are a number of TODOs in the code and I suppose quite a few might be obsolete by now. But I think it would be best if you could take care of cleaning that, you know best which ones to keep 😅